### PR TITLE
Use separate configuration for extracting included proto from dependencies

### PR DESCRIPTION
**Pull request by voidzcy**
Configuration hierarchy for `java` and `java-library` plugins are different:
- In [`java` plugin](https://docs.gradle.org/current/userguide/java_plugin.html#tab:configurations): `compile`(deprecated) is the root configuration, `implementation` extends it and `compileClasspath` extends `implementation`.
- In [`java-library`](https://docs.gradle.org/current/userguide/java_library_plugin.html#sec:java_library_configurations_graph): `api` is the root configuration, `implementation` extends it and `compileClasspath` extends `implementation`.

Currently, our plugin [extracts included proto for compilation from dependencies in `compile` configuration](https://github.com/google/protobuf-gradle-plugin/blob/f91db82a7de0adc1e71774ebeeb2fd1377f6eb0a/src/main/groovy/com/google/protobuf/gradle/ProtobufPlugin.groovy#L369). This is wrong even for `java` plugin. Dependencies added to `implementation` configuration will not be captured by our plugin. If your proto file imports some dependent proto from an `implementation` configuration, you will get an import proto file not found error. Plugin tasks should get dependencies from `compileClasspath` configuration, which inherits all dependencies added to its ancestor configurations. Thus, the plugin is able to capture all dependencies for compilation.

However, extracting included proto from dependencies in `compileClasspath` configuration is not enough if some dependency uses `java-library` plugin. Projects depending on some upstream libraries that use `java-library` plugin will only pull in [classes folder instead of the full JAR for compilation](https://docs.gradle.org/current/userguide/java_library_plugin.html#sec:java_library_classes_usage). So proto files in the upstream libraries will not be extracted and consumer project's compilation will fail if it imports proto files from upstream libraries.

Our solution, as explained in https://github.com/google/protobuf-gradle-plugin/issues/363#issuecomment-562802208 is to create an internal configuration that mirrors `compileClasspath` configuration (which extends `compileOnly` and `implementation`) introduced by `java`/`java-library` plugin and add an attribute to also include `resources` (where the proto files are packaged). We are trying to avoid affecting `java`/`java-library` by modifying `compileClasspath` configuration itself. This new configuration is for our plugin's internal usage and users should not directly add dependencies to it.

Related issues: #363 (most clearly reveals the issue), #242, #338, #258

### CODE REVIEW

File: src/test/groovy/com/google/protobuf/gradle/ProtobufJavaPluginTest.groovy [hunk 1]
```diff
@@ -200,6 +200,62 @@ class ProtobufJavaPluginTest extends Specification {
...
+            .build()
+
+    when: "build is invoked"
+    BuildResult result = GradleRunner.create()
+            .withProjectDir(projectDir)
+            .withArguments('build', '--stacktrace')
+            .withGradleVersion(gradleVersion)
+            .forwardStdOutput(new OutputStreamWriter(System.out))
+            .forwardStdError(new OutputStreamWriter(System.err))
+            .withDebug(true)
```
Comment (ejona86) [inline src/test/groovy/com/google/protobuf/gradle/ProtobufJavaPluginTest.groovy]: Oh, I thought we had deleted all instances of withDebug. We probably want to do that in the future. Don't change anything in this PR, though.

File: gradle/wrapper/gradle-wrapper.properties [hunk 2]
```diff
@@ -2,4 +2,4 @@ distributionBase=GRADLE_USER_HOME
 distributionPath=wrapper/dists
 zipStoreBase=GRADLE_USER_HOME
 zipStorePath=wrapper/dists
-distributionUrl=https\://services.gradle.org/distributions/gradle-5.1-bin.zip
+distributionUrl=https\://services.gradle.org/distributions/gradle-5.6-bin.zip
```
[the remark this answers is not in the document]
Comment (voidzcy) [reply gradle/wrapper/gradle-wrapper.properties]: Gradle 5.6 is required for [`LibraryElements`](https://docs.gradle.org/current/javadoc/org/gradle/api/attributes/LibraryElements.html).

File: src/main/groovy/com/google/protobuf/gradle/ProtobufPlugin.groovy [hunk 3]
```diff
@@ -158,18 +160,37 @@ class ProtobufPlugin implements Plugin<Project> {
...
           extendsFrom = []
         }
       }
+
+      // Create a 'compileProto' configuration as a bucket of dependencies that contains resources
+      // attribute for compileClasspath dependencies. This works around 'java-library' plugin
+      // not exposing resources to consumers for compilation.
+      // Some Android sourceSet does not have 'compileClasspath' configuration, not even
+      // 'compile' or 'implementation'.
+      String compileProtoConfigName = Utils.getConfigName(sourceSetName, 'compileProto')
```
[the remark this answers is not in the document]
Comment (voidzcy) [reply src/main/groovy/com/google/protobuf/gradle/ProtobufPlugin.groovy]: Sure, I had time remembering the name you suggested 😢 when I was writing this ...

File: src/main/groovy/com/google/protobuf/gradle/ProtobufPlugin.groovy [hunk 4]
```diff
@@ -158,18 +160,37 @@ class ProtobufPlugin implements Plugin<Project> {
...
+      String compileClasspathConfigName = Utils.getConfigName(sourceSetName, 'compileClasspath')
+      if (project.configurations.findByName(compileClasspathConfigName) &&
+              project.configurations.findByName(compileProtoConfigName) == null) {
+        project.configurations.create(compileProtoConfigName) {
+          visible = false
+          transitive = true
+          extendsFrom = [project.configurations.findByName(compileClasspathConfigName)]
```
[the remark this answers is not in the document]
Comment (voidzcy) [reply src/main/groovy/com/google/protobuf/gradle/ProtobufPlugin.groovy]: Basically the same. I was thinking about `compileClasspath` is a "more extended" configuration, we would be less likely to miss out something.  Do you think which approach is more reasonable?

[the remark this answers is not in the document]
Comment (ejona86) [reply src/main/groovy/com/google/protobuf/gradle/ProtobufPlugin.groovy]: That "more extended" is what bothers me. I don't really like to be using things that I don't know I'm using, as someone could end up depending on them. But I see danger from both options. I'm most worried about all the strange things a user may change.

[the remark this answers is not in the document]
Comment (voidzcy) [reply src/main/groovy/com/google/protobuf/gradle/ProtobufPlugin.groovy]: Alright, let's be conservative and mirror how Java Base Plugin's tasks getting dependencies. I changed it to extend from `compileOnly` and `implementation`.

**Pull request merged**